### PR TITLE
Merge ipfs+rtmp server into one and drop FFMPEG

### PR DESCRIPTION
**Pull request by darkdrgn2k**
- [x] merge ipfs and rtmp server
- [x] switched to nginx hls
- [x] mismatched if in process stream needs to be fixed
- [x] nginx does not start up
- [x] make sure nginx service is installed properly
- [x] Update documentation to stream key to `stream1`
- [x] Mirror does not start up properly
```
2 error(s) occurred:

* digitalocean_record.ipfs-server-gateway: 1 error(s) occurred:

* digitalocean_record.ipfs-server-gateway: GET https://api.digitalocean.com/v2/domains/live.meshwithme.online/records/68073033: 500 Server was unable to give you a response.
* digitalocean_record.ipfs-server-gateway-private: 1 error(s) occurred:

* digitalocean_record.ipfs-server-gateway-private: GET https://api.digitalocean.com/v2/domains/live.meshwithme.online/records/68073035: 500 Server was unable to give you a response.

```

### CODE REVIEW

**Comment by darkdrgn2k** [comment]
Discuss PROS and CON in this method.

File: README.md [hunk 1]
```diff
@@ -202,7 +202,7 @@ installed on your local machine, which can be the same device running OBS Studio
     To authenticate using OpenVPN, connect with your OpenVPN client using `client.conf` or
     `client.ovpn`, then publish your OBS Studio stream to:
 
-        rtmp://10.10.10.1:1935/live
+        rtmp://10.10.10.1:1935/live/stream1
```
Comment (benhylau) [inline README.md]: Can we choose something other than `stream1`? Is this meant for multi-stream support?

File: terraform/Security [hunk 2]
```diff
@@ -0,0 +1 @@
+Add the SSH key to your Digital Ocean account under Settings
```
Comment (benhylau) [inline terraform/Security]: What is this file?

File: terraform/main.tf [hunk 3]
```diff
@@ -301,29 +258,24 @@ output "dns_records" {
   value      = [
     "                  ${digitalocean_domain.ipfs-live-streaming.name} = ${digitalocean_domain.ipfs-live-streaming.ip_address}",
     "      ${digitalocean_record.rtmp-server.fqdn} = ${digitalocean_record.rtmp-server.value}",
-    "      ${digitalocean_record.ipfs-server.fqdn} = ${digitalocean_record.ipfs-server.value}",
-    "     ${digitalocean_record.ipfs-server-gateway.fqdn} = ${digitalocean_record.ipfs-server-gateway.value}",
+    "      ${digitalocean_record.rtmp-server.fqdn} = ${digitalocean_record.rtmp-server.value}",
```
Comment (benhylau) [inline terraform/main.tf]: You are collapsing both servers into `rtmp-server`, but it still has an ipfs gateway, no?

File: terraform/main.tf [hunk 4]
```diff
@@ -301,29 +258,24 @@ output "dns_records" {
...
     "${zipmap(digitalocean_record.ipfs-mirror-private.*.fqdn, digitalocean_record.ipfs-mirror-private.*.value)}",
     "${zipmap(digitalocean_record.ipfs-mirror-gateway-private.*.fqdn, digitalocean_record.ipfs-mirror-gateway-private.*.value)}",
     "      ${digitalocean_record.rtmp-server-v6.fqdn} = ${digitalocean_record.rtmp-server-v6.value}",
-    "      ${digitalocean_record.ipfs-server-v6.fqdn} = ${digitalocean_record.ipfs-server-v6.value}",
-    "     ${digitalocean_record.ipfs-server-gateway-v6.fqdn} = ${digitalocean_record.ipfs-server-gateway-v6.value}",
     "${zipmap(digitalocean_record.ipfs-mirror-v6.*.fqdn, digitalocean_record.ipfs-mirror-v6.*.value)}",
     "${zipmap(digitalocean_record.ipfs-mirror-gateway-v6.*.fqdn, digitalocean_record.ipfs-mirror-gateway-v6.*.value)}",
     "     ${digitalocean_record.publish-openvpn.fqdn} = ${digitalocean_record.publish-openvpn.value}",
-    "   ${digitalocean_record.publish-yggdrasil.fqdn} = ${digitalocean_record.publish-yggdrasil.value}",
+    "     ${digitalocean_record.publish-yggdrasil.fqdn} = ${digitalocean_record.publish-yggdrasil.value}",
```
Comment (benhylau) [inline terraform/main.tf]: Why this space change?

File: terraform/rtmp-server/bootstrap-post-dns.sh [hunk 5]
```diff
@@ -17,15 +17,14 @@ systemctl stop nginx.service
 certbot certonly -n --agree-tos --standalone --email "${EMAIL_ADDRESS}" -d "${DOMAIN_NAME}" -d "ipfs-server.${DOMAIN_NAME}" -d "ipfs-gateway.${DOMAIN_NAME}"
 
 # Configure nginx with HTTPS
-cp -f /tmp/ipfs-server/nginx-default /etc/nginx/sites-available/default
-sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g" /etc/nginx/sites-available/default
+cp -f /tmp/rtmp-server/nginx-default  /usr/local/nginx/conf/conf.d/default
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap-post-dns.sh]:
```suggestion
cp -f /tmp/rtmp-server/nginx-default /usr/local/nginx/conf/conf.d/default
```

File: terraform/rtmp-server/bootstrap-post-dns.sh [hunk 6]
```diff
@@ -17,15 +17,14 @@ systemctl stop nginx.service
...
+sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g"  /usr/local/nginx/conf/conf.d/default
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap-post-dns.sh]:
```suggestion
sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g" /usr/local/nginx/conf/conf.d/default
```

File: terraform/rtmp-server/bootstrap-post-dns.sh [hunk 7]
```diff
@@ -17,15 +17,14 @@ systemctl stop nginx.service
...
+cp -f /tmp/rtmp-server/nginx-gateway "/usr/local/nginx/conf/conf.d/ipfs-gateway.${DOMAIN_NAME}"
+sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g" "/usr/local/nginx/conf/conf.d/ipfs-gateway.${DOMAIN_NAME}"
 
 # Configure auto-renewals
 echo "30 2 * * 1 certbot renew >> /var/log/letsencrypt/letsencrypt.log" >> /etc/crontab
-echo "35 2 * * 1 systemctl reload nginx" >> /etc/crontab
\ No newline at end of file
+echo "35 2 * * 1 systemctl reload nginx" >> /etc/crontab
+
+systemctl start nginx.service
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap-post-dns.sh]: Why is this moved?

File: terraform/rtmp-server/bootstrap-post-dns.sh [hunk 8]
```diff
@@ -17,15 +17,14 @@ systemctl stop nginx.service
...
-sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g" /etc/nginx/sites-available/default
+cp -f /tmp/rtmp-server/nginx-default  /usr/local/nginx/conf/conf.d/default
+sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g"  /usr/local/nginx/conf/conf.d/default
 
-cp -f /tmp/ipfs-server/nginx-gateway "/etc/nginx/sites-available/ipfs-gateway.${DOMAIN_NAME}"
-sed -i "s#__DOMAIN_NAME__#${DOMAIN_NAME}#g" "/etc/nginx/sites-available/ipfs-gateway.${DOMAIN_NAME}"
-ln -s "/etc/nginx/sites-available/ipfs-gateway.${DOMAIN_NAME}" "/etc/nginx/sites-enabled/ipfs-gateway.${DOMAIN_NAME}"
-
-systemctl start nginx.service
+cp -f /tmp/rtmp-server/nginx-gateway "/usr/local/nginx/conf/conf.d/ipfs-gateway.${DOMAIN_NAME}"
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap-post-dns.sh]: We need to print the URL of this gateway in terraform provision complete

File: terraform/rtmp-server/bootstrap.sh [hunk 9]
```diff
@@ -27,7 +34,13 @@ apt install -y \
   libssl-dev \
   zlibc \
   zlib1g \
-  zlib1g-dev
+  zlib1g-dev \
+  bc \
+  certbot \
+  jq
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap.sh]: These are sorted alphabetically

File: terraform/rtmp-server/bootstrap.sh [hunk 10]
```diff
@@ -41,7 +54,6 @@ tar -C /usr/local -xzf /tmp/golang/go1.11.2.linux-amd64.tar.gz
 . /etc/profile
 
 # Create directory for generating client keys
-mkdir /root/client-keys
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap.sh]: If you delete a line of code, please make sure to address the associated comments and blank lines

File: terraform/rtmp-server/bootstrap.sh [hunk 11]
```diff
@@ -193,5 +205,82 @@ mkdir /root/hls
...
+until [[ `ipfs id >/dev/null 2>&1; echo $?` -eq 0 ]]; do
+  sleep 1
+done
+sleep 10
+
+# Write IPFS identity to client file
+IPFS_ID=`ipfs id | jq .ID | sed 's/"//g'`
+echo -n "$IPFS_ID" > ~/client-keys/ipfs_id
+
+
```
Comment (benhylau) [inline terraform/rtmp-server/bootstrap.sh]:
```suggestion
```

File: terraform/rtmp-server/nginx.service [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# Stop dance for nginx
```
Comment (benhylau) [inline terraform/rtmp-server/nginx.service]: Is this a backport from later versions of nginx?

File: terraform/rtmp-server/process-stream.sh [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+# Load settings
+. ~/settings
+
+# Create directory for HLS content
+rm -rf ~/hls/*
+cd ~/hls
+
+DISCONNET=1
+what="stream1"
```
Comment (benhylau) [inline terraform/rtmp-server/process-stream.sh]: Change name?

File: README.md (hunk 1 above)
Comment (darkdrgn2k) [reply README.md]: You can choose anything you want.  

Support for multiple streams is trivial at this point.  Right now process stream is locked on to process "stream1.m3u8"  however it can easily either be spawned as a parallel process for each new stream, or simply written to look for streamN.m3u8 and place the fines in the respective liveN.m3u8 file.

This was a possibility for multiple rooms at Our Networks, however as it seems its not required its not been completed.

Issue in https://github.com/tomeshnet/ipfs-live-streaming/issues/79

File: terraform/Security (hunk 2 above)
Comment (darkdrgn2k) [reply terraform/Security]: I have no idea how it got there.  Removed.

File: terraform/rtmp-server/nginx.service (hunk 12 above)
Comment (darkdrgn2k) [reply terraform/rtmp-server/nginx.service]: I THINK it was taken from the service file in Raspbian GNU/Linux 8.
Previous versions of this server spawned nginx as a process durring startup, not a restartable service.

File: terraform/rtmp-server/bootstrap-post-dns.sh (hunk 7 above)
Comment (darkdrgn2k) [reply terraform/rtmp-server/bootstrap-post-dns.sh]: Just added an nginx start, since we are not starting it as a process

File: terraform/rtmp-server/bootstrap-post-dns.sh (hunk 8 above)
Comment (darkdrgn2k) [reply terraform/rtmp-server/bootstrap-post-dns.sh]: Issue #91
